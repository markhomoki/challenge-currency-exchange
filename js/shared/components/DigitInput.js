import React from 'react';
import PropTypes from 'prop-types';

export default class DigitInput extends React.PureComponent {

	static propTypes = {
		direction: PropTypes.oneOf(['from', 'to']).isRequired,
		onInputChange: PropTypes.func,
		value: PropTypes.oneOfType([
			PropTypes.string,
			PropTypes.number,
		]),
	}

	static defaultProps = {
		value: '',
	}

	static getDerivedStateFromProps(nextProps, prevState) {

		if (nextProps.value != prevState.value) {
			const newValue = nextProps.value ? nextProps.value.toString() : '';
			return { value: newValue };
		}

		return false;
	}

	constructor(props) {
		super(props);

		this.state = {
			value: this.props.value,
		};
	}

	handleChange = (e) => {
		const { value } = e.target;

		if (value.match(/[a-z]/i) || value === '0') {
			return;
		}

		this.updateValue(value);
	}

	updateValue = (value) => {
		const { onInputChange } = this.props;
		const formattedValue = value.replace(/\+|-/g, '');
		const floatedValue = formattedValue ? parseFloat(formattedValue) : 0;

		this.setState({ value: formattedValue });

		if (onInputChange) {
			onInputChange(floatedValue, this.props.direction);
		}
	}

	render() {
		const {
			direction,
		} = this.props;

		const {
			value,
		} = this.state;

		const prefix = direction === 'from' ? '-' : '+';
		const formattedValue = value.length > 0 ? `${prefix}${value}` : '';

		return (
			<div className="digit-input-wrapper">
				<input
					type="text"
					className="digit-input"
					onChange={this.handleChange}
					value={formattedValue}
					placeholder="0"
				/>
			</div>
		);
	}

}
