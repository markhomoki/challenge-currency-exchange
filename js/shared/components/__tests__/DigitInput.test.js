import React from 'react';
import renderer from 'react-test-renderer';
import DigitInput from '../DigitInput';

describe('DigitInput', () => {
	let wrapper;
	const element = (
		<DigitInput direction="from" />
	);

	beforeEach(() => {
		wrapper = mount(element);
	});

	it('renders correctly', () => {
		const tree = renderer.create(element).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('doesn\'t accept letter', () => {
		wrapper.instance().handleChange({ target: { value: 'Hello World' } });
		expect(wrapper.state().value).toEqual('');
	});

	it('doesn\'t accept leading 0', () => {
		wrapper.instance().handleChange({ target: { value: '0' } });
		expect(wrapper.state().value).toEqual('');
	});
});
