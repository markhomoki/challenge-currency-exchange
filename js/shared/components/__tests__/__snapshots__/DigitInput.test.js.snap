// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DigitInput renders correctly 1`] = `
<div
  className="digit-input-wrapper"
>
  <input
    className="digit-input"
    onChange={[Function]}
    placeholder="0"
    type="text"
    value=""
  />
</div>
`;
